shared/shared/postgresql_backend/pg_connection_handler.py

```diff
@@ -30,83 +33,81 @@ def __init__(self, connection_handler):
         self.connection_handler = connection_handler
 
     def __enter__(self):
-        self.connection_handler.connection.__enter__()
-        self.connection_handler.set_transaction_running(True)
+        self.connection = self.connection_handler.get_connection()
+        self.connection.__enter__()
 
     def __exit__(self, type, value, traceback):
-        self.connection_handler.set_transaction_running(False)
-        if self.connection_handler.connection:
-            self.connection_handler.connection.__exit__(type, value, traceback)
+        self.connection.__exit__(type, value, traceback)
+        self.connection_handler.put_connection(self.connection)
```

Comment:
Can the connection_pool handle errored(/invalid?) connections. Please test a connection pool with maxconn=1: Get a connection, do something that produces an error (e.g. invalid statements) and put it back. Then get a connection from the pool. This connection should be ok to use.

Reply:
(Supposedly, psycopg2 takes care of that internally.)[https://github.com/psycopg/psycopg2/issues/62] I don't think we need to test this explicitly then but should instead rely on psycopg2. After all, we don't test that flask/gunicorn serve the route they're supposed to either.